### PR TITLE
[FLINK-28710][table-planner] Supports dynamic filtering execution

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*Supports dynamic filtering execution*

## Brief change log

  - *The fact source of DPP must be FLIP-27 source*
  - *Supports dynamic filtering execution*

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

This change added tests and can be verified as follows:

  - *Extended DynamicPartitionPruningRuleTest  to verify the legacy source can't be the fact source for DPP query*
  - *Added DynamicFilteringTest to verify the exec plan for DPP query*
  - *Added DynamicFilteringITCase to verify the execution result for DPP query*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by godfreyhe** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TestValuesTableFactory.java [hunk 1]
```diff
@@ -960,11 +997,12 @@ protected Collection<RowData> convertToRowData(DataStructureConverter converter)
                     final RowData rowData = (RowData) converter.toInternal(row);
                     if (rowData != null) {
                         rowData.setRowKind(row.getKind());
-                        result.add(rowData);
+                        partitionResult.add(rowData);
+                        size++;
```
Comment (gaoyunhaii) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TestValuesTableFactory.java:1001]: It seems size == partitionResult.size() ? Could we directly use partitionResult.size() ?

**Review by gaoyunhaii — COMMENTED** [review]
Thanks @godfreyhe for the PR! I have left some comments.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/connector/source/ValuesSourceReader.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+import java.io.ByteArrayInputStream;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+
+/** A {@link SourceReader} implementation that reads data from a list. */
+public class ValuesSourceReader implements SourceReader<RowData, ValuesSourceSplit> {
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/connector/source/ValuesSourceReader.java:44]: make sense too me. There are also many source readers that need to be refactored, such as: IteratorSourceReader,  FromElementsSourceReader. We can do it in another PR

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/connector/source/enumerator/DynamicFilteringValuesSourceEnumerator.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    public void addSplitsBack(List<ValuesSourcePartitionSplit> splits, int subtaskId) {
+        remainingSplits.addAll(splits);
+    }
+
+    @Override
+    public void addReader(int subtaskId) {}
+
+    @Override
+    public NoOpEnumState snapshotState(long checkpointId) throws Exception {
+        return null;
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/connector/source/enumerator/DynamicFilteringValuesSourceEnumerator.java:144]: good catch

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TestValuesTableFactory.java (hunk 1 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/factories/TestValuesTableFactory.java:1001]: The `partitionResult.size()`  is the size of a specific partition, while the `size` is all accepted partitions total size. There is a for-loop at line 961

**Comment by godfreyhe** [comment]
Thanks for the review @gaoyunhaii , I have updated the pr

**Review by gaoyunhaii — COMMENTED** [review]
Thanks @godfreyhe for the update! I only have some minor comments.

**Comment by godfreyhe** [comment]
@flinkbot run azure

**Comment by godfreyhe** [comment]
The failure is due to the existed issue https://issues.apache.org/jira/browse/FLINK-28844

**Review by gaoyunhaii — APPROVED** [review]
Thanks @godfreyhe for the PR! LGTM

**Pull request closed without merging**